exchange/tmax_adjustments.go

```diff
@@ -46,6 +47,9 @@ func (b *bidderTmaxCtx) Deadline() (deadline time.Time, ok bool) {
 	deadline, ok = b.ctx.Deadline()
 	return
 }
+func (b *bidderTmaxCtx) Until(t time.Time) time.Duration {
+	return time.Until(t)
```

Comment:
Nitpick: Are we planning to add more functionality into `Until` in the future? If not, given that it doesn't access nor works with `bidderTmaxCtx`, maybe we can get rid of it and simply use the `time` library function. `"time"` is already listed in the imports of `exchange/bidder.go` so we wouldn't need the extra import either.
```
741   func hasShorterDurationThanTmax(ctx bidderTmaxContext, tmaxAdjustments TmaxAdjustmentsPreprocessed) bool {
742       if tmaxAdjustments.IsEnforced {
743           if deadline, ok := ctx.Deadline(); ok {
744               overheadNS := time.Duration(tmaxAdjustments.BidderNetworkLatencyBuffer+tmaxAdjustments.PBSResponsePreparationDuration) * time.Millisecond
745               bidderTmax := deadline.Add(-overheadNS)
746  
747 -             remainingDuration := ctx.Until(bidderTmax).Milliseconds()
    +             remainingDuration := time.Until(bidderTmax).Milliseconds()
748               return remainingDuration < int64(tmaxAdjustments.BidderResponseDurationMin)
749           }
750       }
751       return false
752   }
exchange/bidder.go
```
This could also be the case of `RemainingDurationMS`. Unless this is for testing purposes and I'm missing the big picture.

Reply:
`time.Until()` internally calls `time.Now()`.  Its difficult to test code path when value returned from `time.Now()` is used. By using `ctx.Until()` we can mock this return value.  

Refer [TestHasShorterDurationThanTmax](https://github.com/prebid/prebid-server/blob/1b6ebea1dd504584df7f8000ee3bb9a0b447d727/exchange/bidder_test.go#L3150) unit test. 
https://github.com/prebid/prebid-server/blob/1b6ebea1dd504584df7f8000ee3bb9a0b447d727/exchange/bidder_test.go#L3156

In the test, `mockBidderTmaxCtx` is passed as implementer for `biddetTmaxCtx`. So during unit test `mockBidderTmaxCtx.Until()` is called. 
https://github.com/prebid/prebid-server/blob/1b6ebea1dd504584df7f8000ee3bb9a0b447d727/exchange/bidder_test.go#L3081-L3083

By using `ctx.Until()` with `mockBidderTmaxCtx`, we can easily control the return value for testing purposes. This approach enables us to test specific code paths that were previously difficult `time.Until()`.

Reply:
How about avoiding `time.Now()` entirely by storing the `openrtb2.BidRequest.TMax` field in the `tmaxAdjustments` parameter and using it for the calculation. I believe both calculations and test cases could be simplified:
```
741 - func hasShorterDurationThanTmax(ctx bidderTmaxContext, tmaxAdjustments TmaxAdjustmentsPreprocessed) bool {
    + // hasShorterDurationThanTmax returns true if there's enough tmax to make the roundtrip to the bid server 
    + // doing a calculation that involves the original bid request tmax
    + func hasShorterDurationThanTmax(tmaxAdjustments TmaxAdjustmentsPreprocessed) bool {
742       if tmaxAdjustments.IsEnforced {
743 -         if deadline, ok := ctx.Deadline(); ok {
744               overheadNS := time.Duration(tmaxAdjustments.BidderNetworkLatencyBuffer+tmaxAdjustments.PBSResponsePreparationDuration) * time.Millisecond
745 -             bidderTmax := deadline.Add(-overheadNS)
746
747 -             remainingDuration := ctx.Until(bidderTmax).Milliseconds()
    +             remainingDuration := tmaxAdjustments.BidRequestTmax - overheadNS
748               return remainingDuration < int64(tmaxAdjustments.BidderResponseDurationMin)
749 -         }
750       }
751       return false
752   }
exchange/bidder.go
```
We would have to copy the `openrtb2.BidRequest.TMax` into `tmaxAdjustments.BidRequestTmax` and recalculate it's value inside `func (e *exchange) makeAuctionContext(ctx context.Context, needsCache bool)` (`exchange/exchange.go` line 646) if necessary.

Do you think we can pull this off?

Reply:
We need to get the elapsed time since start of auction handler function. 
For this we are using, deadline context intialised at the start. Refer following:

https://github.com/prebid/prebid-server/blob/35db00f170625a2436019a0a175864ed6f2d437e/endpoints/openrtb2/auction.go#L208-L213

This elapsed time from deadline context is fetched using ` ctx.Deadline()` and then `BidderNetworkLatencyBuffer`, `PBSResponsePreparationDuration` is deducted from resultant deadline to get bidderTmax.

I am not sure how using `openrtb2.BidRequest.TMax` we can get elapsed time? may be I am missing something